### PR TITLE
Add SpeedController nodes to adjust replanning rate according to speed

### PR DESCRIPTION
**Pull request by naiveHobo**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1740 |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | Gazebo simulation of TurtleBot3 |

---

## Description of contribution in a few bullet points

* Added ```SpeedController``` decorator and condition nodes
* Updated default BTs to use ```SpeedController``` instead of ```RateController```

## Description of documentation updates required from your changes

* Update [navigation.ros.org](https://navigation.ros.org/)

### CODE REVIEW

**Comment by naiveHobo** [comment]
Can't think of a proper name for an analor condition node for the ```SpeedController``` decorator. It's a condition which returns success at a rate adjusted according to the current speed. What do we call such a condition? ```ScaledTimeExpired```?

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+SpeedController::SpeedController(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf),
+  first_time_(false),
+  period_(1.0)
+{
+  getInput("min_rate", min_rate_);
+  getInput("max_rate", max_rate_);
+  getInput("min_speed", min_speed_);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.cpp:34]: Wouldn't the minimum just be 0? That does ask the question: what if this is negative for an extended duration?

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  getInput("max_speed", max_speed_);
+
+  d_rate_ = max_rate_ - min_rate_;
+  d_speed_ = max_speed_ - min_speed_;
+
+  node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+
+  odom_sub_ = node_->create_subscription<nav_msgs::msg::Odometry>(
+    "odom",
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.cpp:43]: Topic should be configurable through inputs as well

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    rclcpp::SystemDefaultsQoS(),
+    std::bind(&SpeedController::onOdomReceived, this, std::placeholders::_1));
+}
+
+inline BT::NodeStatus SpeedController::tick()
+{
+  if (status() == BT::NodeStatus::IDLE) {
+    // Reset the starting position since we're starting a new iteration of
+    // the distance controller (moving from IDLE to RUNNING)
+    start_ = std::chrono::steady_clock::now();
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.cpp:53]: Use ROS time please

File: nav2_bt_navigator/behavior_trees/navigate_w_replanning_and_recovery.xml [hunk 4]
```diff
@@ -7,12 +7,12 @@
   <BehaviorTree ID="MainTree">
     <RecoveryNode number_of_retries="6" name="NavigateRecovery">
       <PipelineSequence name="NavigateWithReplanning">
-        <RateController hz="1.0">
+        <SpeedController min_rate="0.1" max_rate="1.0" min_speed="0.0" max_speed="0.25">
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/navigate_w_replanning_and_recovery.xml:10]: I don't think we should be updating these BTs with this. Lets just add the controller and we can discuss the right default trees / nodes we want to use in another ticket.

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 5]
```diff
@@ -49,6 +49,7 @@ BtNavigator::BtNavigator()
     "nav2_reinitialize_global_localization_service_bt_node",
     "nav2_rate_controller_bt_node",
     "nav2_distance_controller_bt_node",
+    "nav2_speed_controller_bt_node",
```
Comment (SteveMacenski) [inline nav2_bt_navigator/src/bt_navigator.cpp:52]: Please submit a PR to add this PR to migration guide + this BT node to the plugins list.

File: nav2_behavior_tree/plugins/decorator/speed_controller.hpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    };
+  }
+
+private:
+  BT::NodeStatus tick() override;
+
+  void onOdomReceived(nav_msgs::msg::Odometry::SharedPtr msg);
+
+  // Scale the rate based speed
+  inline double getScaledRate(double speed)
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.hpp:54]:
```suggestion
  inline double getScaledRate(const double & speed)
```

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+
+  return status();
+}
+
+void SpeedController::onOdomReceived(const nav_msgs::msg::Odometry::SharedPtr msg)
+{
+  // Calculate speed using velocity components
+  double speed = std::hypot(msg->twist.twist.linear.x, msg->twist.twist.linear.y);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.cpp:95]: You need to check that this is non-zero or you'll get nans

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp (hunk 1 above)
Comment (naiveHobo) [reply nav2_behavior_tree/plugins/decorator/speed_controller.cpp:34]: I added the minimum speed param as a way to clip the rate below a certain traveling speed. For example, if min_speed was set to 0.5 m/s, the rate will be clamped to min_rate at any speed <=0.5

Comment (SteveMacenski) [reply nav2_behavior_tree/plugins/decorator/speed_controller.cpp:34]: Got it, good idea.

Comment (SteveMacenski) [reply nav2_behavior_tree/plugins/decorator/speed_controller.cpp:34]: I think you should expose these ports in the BT when we use it. I want people to specifically see what the min/max speed/rates are. If a robot goes any normal speed (e.g. non-turtlebot) 0.25 is nothing. Maybe bump that default max to 0.5 m/s.

**Comment by naiveHobo** [comment]
> I think we need to find a way to smooth out the results a little, maybe using a 3rd order rolling filter just to average out some results to have a smoother transition in periods.

That sounds like a good general addition to the stack. I see a lot of TODOs in [`is_stuck_condition`](https://github.com/ros-planning/navigation2/blob/master/nav2_behavior_tree/plugins/condition/is_stuck_condition.cpp) calling for a similar thing.
https://github.com/ros-planning/navigation2/blob/ca33b19bfb75dbdda5929a299ab2baa8ef76ed1a/nav2_behavior_tree/plugins/condition/is_stuck_condition.cpp#L109

Do you think the should be done in a `RobotState` class? Among other things, the class can subscribe to the odometry and provide methods to get either the raw current velocity or a smooth velocity signal using a filtering technique.

**Comment by naiveHobo** [comment]
After adding a simple moving average, here's how the speed vs rate of replanning plots look:

**Filter size**: 10
Figure_0

**Filter size**: 20
Figure_1


**Filter size**: 50
Figure_3


**Filter size**: 100
Figure_2

**Comment by SteveMacenski** [comment]
What rate was the odometry being published at? 

Maybe the parameter might make sense as a duration rather than a number of samples since that is more portable to many robots vs. # samples that will change for every robot.

**Comment by naiveHobo** [comment]
> What rate was the odometry being published at?

About 29hz on average.

> Maybe the parameter might make sense as a duration rather than a number of samples since that is more portable to many robots vs. # samples that will change for every robot. Or offering both options in nav2_util.

That's a good suggestion, time based averaging would be better. I'll add that.

**Comment by SteveMacenski** [comment]
I would have thought something like 100ms duration, but that seems pretty short with only 1-10 samples for many robots. 300ms would be 9 samples on this example, which sounds as high as I think is reasonable. What do you think?

**Comment by naiveHobo** [comment]
Let's keep the default to 500ms I think, that should range from about 10-15hz in most cases.

> When you update the period, should that period apply for the next cycle and then we update the period again, or should the period be constantly updating even in the middle of a cycle?

Bit confused about what direction to go here. Right now, the period is updated on every tick and is applicable only to that particular tick. What do you suggest?

**Comment by SteveMacenski** [comment]
500ms seems high if someone wants to have their replanning rate at 2-5hz, which I've certainly seen. 

> Right now, the period is updated on every tick and is applicable only to that particular tick.

I think that's a little too unstable. I think it should be that a period is selected (initialize at the high end of the range), used, and once it triggers a replan or is reset by `IDLE`, a new period is computed using the current state of the odometry filter. Having a moving target in the middle of a sleep seems like something hard to debug problems with and potentially harmful if in a very large range of speeds

**Comment by naiveHobo** [comment]
> and once it triggers a replan or is reset by IDLE, a new period is computed using the current state of the odometry filter.

Replanning is triggered on the very first tick when the robot is stationary, which would immediately set the period to the maximum value (least rate) after the first plan. With the minimum rate of 0.1hz that I've been using, the second time replanning will be triggered would be after 10 seconds. Similar problem with IDLE, the robot is most likely stationary when the node is IDLE.

Figure_1

---

Instead, how about the period is set to the least value (maximum rate) when the node is IDLE and a new period is computed using the current state of the odometry filter only when the last period has been exceeded. This is how the plot looks with this configuration:

Figure_0

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp (hunk 7 above)
Comment (naiveHobo) [reply nav2_behavior_tree/plugins/decorator/speed_controller.cpp]: I'm not sure this is true, `std::hypot` just return 0 if either of the values is 0.

Comment (SteveMacenski) [reply nav2_behavior_tree/plugins/decorator/speed_controller.cpp]: @naiveHobo I think this highlighted the wrong line on speed, I meant the line after this one. If you have speed = 0, then `period_ = 1/speed` is `1/0` which will be a nan.

File: nav2_behavior_tree/plugins/decorator/speed_controller.hpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        (((speed - min_speed_) / d_speed_) * d_rate_) + min_rate_,
+        max_rate_), min_rate_);
+  }
+
+  // Update period based on current smoothed speed and reset timer
+  inline void updatePeriod()
+  {
+    auto velocity = odom_smoother_->getTwist();
+    double speed = std::hypot(velocity.linear.x, velocity.linear.y);
+    period_ = 1.0 / getScaledRate(speed);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.hpp:68]: `1.0 / getScaledRate(speed);` can be a nan

File: nav2_behavior_tree/README.md [hunk 9]
```diff
@@ -75,6 +75,7 @@ The nav2_behavior_tree package provides several navigation-specific nodes that a
 | NavigateToPose | Action | Invokes the NavigateToPose ROS2 action server, which is implemented by the bt_navigator module. |
 | RateController | Decorator | A node that throttles the tick rate for its child. The tick rate can be supplied to the node as a parameter. The node returns RUNNING when it is not ticking its child. Currently, in the navigation stack, the `RateController` is used to adjust the rate at which the `ComputePathToPose` and `GoalReached` nodes are ticked. |
 | DistanceController | Decorator | A node that controls the tick rate for its child based on the distance traveled. The distance to be traveled before replanning can be supplied to the node as a parameter. The node returns RUNNING when it is not ticking its child. Currently, in the navigation stack, the `DistanceController` is used to adjust the rate at which the `ComputePathToPose` and `GoalReached` nodes are ticked. |
+| SpeedController | Decorator | A node that controls the tick rate for its child based on the current speed. This decorator offers the most flexibility as the user can set the minimum/maximum tick rate which is adjusted according to the current speed. |
```
Comment (SteveMacenski) [inline nav2_behavior_tree/README.md:78]: "current robot speed", lets make it specific, given the "speed" could refer to other things in this context.

File: nav2_util/src/odometry_utils.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+{
+
+OdomSmoother::OdomSmoother(
+  rclcpp::Node::SharedPtr nh,
+  double filter_duration)
+: node_(nh),
+  odom_history_duration_(rclcpp::Duration::from_seconds(filter_duration))
+{
+  std::string odom_topic;
+  nh->get_parameter_or("odom_topic", odom_topic, std::string("odom"));
```
Comment (SteveMacenski) [inline nav2_util/src/odometry_utils.cpp:33]: This parameter is never declared, so it will always fail to change the topic. Because this is a utility, my recommendation is making this a constructor parameter for the user of this object to provide and set the topic as a parameter in the bt navigator and declare it (and add to yaml).  Then you can get in the speed controller and pass to this utility.

**Comment by SteveMacenski** [comment]
> Replanning is triggered on the very first tick when the robot is stationary, which would immediately set the period to the maximum value (least rate) after the first plan

I would like the rate to start off with the fastest replanning, not the slowest. E.g. start off with 10hz not 0.1hz. In case this is set to be really, really low, we don't want issues where the robot can't get out of the gate due to high traffic in the robot's dock area. I think conservatively on startup, we want to plan more than less. 

> EDIT: Those sharp jumps in the rate from 0.1hz to 1hz in the second graph is because of setting the period to least (maximum rate) on IDLE. One way to reduce the amplitude of those jumps is to set the period to the mean value on IDLE (period = 2.0 / (max_rate + min_rate)). This is how it looks (best plot till now in my opinion):

Agree, that last is best. When is it `IDLE` for that to be the behavior? Shouldn't this always be in a `RUNNING` state except for startup or canceling goals? I suppose it return success after a replan, so maybe its idle for just 0.1s until a new tick on the BT happens, but it seems like the replanning rate shouldn't rely on the state of the BT node. Because the state changes SUCCESS and IDLE relate to the fact that it replanned after a duration, it should then use the new period to replan not a hardcoded value, don't you think?

File: nav2_behavior_tree/plugins/decorator/speed_controller.hpp (hunk 8 above)
Comment (naiveHobo) [reply nav2_behavior_tree/plugins/decorator/speed_controller.hpp:68]: The rate returned by `getScaledRate` should be in the range [min_rate, max_rate] which is taken as a BT node input by the user. So the better way would be to set the `min_rate` to some default small value (like 0.01hz) if the `min_rate` provided by the user is <= 0?

**Comment by naiveHobo** [comment]
> I think algorithmically the smoother doesn't do what you think it does because the history isn't a fixed size after startup.

I don't think this is correct. The algorithm maintains a cumulated velocity where it adds or subtracts the velocity values while popping or pushing back a message to the history. After the history is updated, the size of the history is used to average out the cumulated velocity. It should be more clear if you look at the updated 'test_odometry_utils.cpp`, I've added the exact scenario you mentioned in your comment.

> When is it IDLE for that to be the behavior?

The node is `IDLE` till a new goal is received. Which means that the node should be `IDLE` only on the first tick after a new goal and should be in the other states otherwise. Which is why setting the rate to some hardcoded value (in this case `2.0 / (min_rate + max_rate)`) makes sense in my opinion.

> I would like the rate to start off with the fastest replanning, not the slowest. E.g. start off with 10hz not 0.1hz.

The problem with setting the rate to the fastest replanning limit when the node is `IDLE` could be a problem if the robot takes some time to accelerate (which is what I noticed in the turtlebot3 simulation). On the very first tick, the rate is set to 10hz which means that the next time the rate would be updated is after 100 milliseconds. The speed of the robot in those first 100ms after a new goal is provided would be very close to 0 and the rate would be set close to `min_rate`. So the next replan will only be triggered after 10 seconds.

Setting the rate to `2.0 / (min_rate + max_rate)` is giving the robot enough time to gain some speed after the first replan is triggered which means that when the rate is updated the next time it would be a value closer to `max_rate`.

Does that make sense?

File: nav2_behavior_tree/plugins/decorator/speed_controller.hpp (hunk 8 above)
Comment (SteveMacenski) [reply nav2_behavior_tree/plugins/decorator/speed_controller.hpp]: I think if the user tries to use <= 0 you should actually throw an exception and bring down the tree, that would be a critical failure. Do a log lethal that you can't do a speed controller with no rate because it would be infinite then throw it.

File: nav2_bringup/bringup/params/nav2_params.yaml [hunk 11]
```diff
@@ -51,6 +51,7 @@ bt_navigator:
     use_sim_time: True
     global_frame: map
     robot_base_frame: base_link
+    odom_topic: /odom
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params.yaml:54]: Is that leading slash actually required?

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp [hunk 12]
```diff
@@ -30,19 +30,28 @@ SpeedController::SpeedController(
...
   getInput("min_rate", min_rate_);
   getInput("max_rate", max_rate_);
   getInput("min_speed", min_speed_);
   getInput("max_speed", max_speed_);
 
+  if (min_rate_ <= 0) {
+    RCLCPP_WARN(
+      node_->get_logger(),
+      "min_rate cannot be <= 0, setting to a default value of 0.01 hz");
+    min_rate_ = 0.01;
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.cpp:44]: I think this should be a log fatal and then throw an exception. This is an impossible configuration.

**Comment by SteveMacenski** [comment]
@naiveHobo, where's the error in my math than that the averager works? 

> The node is IDLE till a new goal is received. Which means that the node should be IDLE only on the first tick after a new goal and should be in the other states otherwise. Which is why setting the rate to some hardcoded value (in this case 2.0 / (min_rate + max_rate)) makes sense in my opinion.

Ah, so between goals its always in the running state, regardless if the period is up and it triggers its child? That's not what this says: https://github.com/ros-planning/navigation2/pull/1744/files#diff-25e89abadc3024976ce451151aea91ebR87-R98. This says that when it replans we return success/failure, and if that subtree is preempted a halt will be called back to `IDLE`. 

I'm not sure I buy the logic for why average the max/min, but I think in concept that's fine. As long as its not the lowest, I'm thinking if you have a large range and the lower end of it is really, really low that would not be good. I don't the acceleration of the robot matters. This is just a path not a trajectory, updating soon after should be OK because not much will have changed so it should be a very similar path. Even if its not, you'd rather change quickly before getting up to speed than after you're already up to speed to then have to turn around.

File: nav2_bringup/bringup/params/nav2_params.yaml (hunk 11 above)
Comment (naiveHobo) [reply nav2_bringup/bringup/params/nav2_params.yaml:54]: I don't think so but that's how the other topics were set in the params file, added it to maintain the symmetry

File: nav2_behavior_tree/plugins/decorator/speed_controller.cpp [hunk 13]
```diff
@@ -37,11 +37,10 @@ SpeedController::SpeedController(
   getInput("min_speed", min_speed_);
   getInput("max_speed", max_speed_);
 
-  if (min_rate_ <= 0) {
-    RCLCPP_WARN(
-      node_->get_logger(),
-      "min_rate cannot be <= 0, setting to a default value of 0.01 hz");
-    min_rate_ = 0.01;
+  if (min_rate_ <= 0 || max_rate_ <= 0) {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/decorator/speed_controller.cpp:40]:
```suggestion
  if (min_rate_ <= 0.0 || max_rate_ <= 0.0) {
```

File: nav2_bringup/bringup/params/nav2_params.yaml (hunk 11 above)
Comment (SteveMacenski) [reply nav2_bringup/bringup/params/nav2_params.yaml:54]: OK sounds good

**Comment by naiveHobo** [comment]
> However, if we do 1 then 2 then 3 with this logic we get 0+1/1 = 1 for the first update, (1+2)/2 = 3/2 for the second and (3/2 + 3)/3 = 1.5. 1.5 != 2

On the first update, the history should be [1] and cumulated is also 1, so the returned value will be 1.
On the second update (assuming the stamps are within the duration limit), the history will be [1, 2], cumulated will be 1 + 2 = 3 and returned value will be 3/2 = 1.5.
On the third update, history is [1, 2, 3], cumulated is 3 + 3 = 6, and returned value is 6/3 = 2.

If the third update has a timestamp such that the duration of the full history is exceeded, on the third update the history will be [3] and cumulated will be 3, and returned value will be 3.

To add another case, if the third update exceeds the duration from the 1st update but not the 2nd update, the history will be [2, 3], cumulated will be 5, and returned value will be 5/2 = 2.5

> Ah, so between goals its always in the running state, regardless if the period is up and it triggers its child?

Let me explain the behavior of this node more clearly (I'll update the description as well). 

1. On every tick the status of this node is set to `RUNNING` (https://github.com/ros-planning/navigation2/pull/1744/files#diff-25e89abadc3024976ce451151aea91ebR67)
2. On a particular tick, the planning subtree is executed if either of the following three conditions is true:
  - If it's the first tick after a new goal arrives. 
  - If the previous execution of the planning subtree returned `RUNNING` (so we can check if the planning subtree finished execution)
  - If the current period was exceeded

---

If the robot has just been launched or the last goal was reached, the status of `SpeedController` is IDLE. This is when we are hardcoding the period to some value (either max rate or avg. of min/max rate).

In the second to last graph I put up earlier, the period was set to 100 milliseconds (rate = 10hz) on IDLE. Here's a graph explaining the sharp jumps in that graph:

quick

**Comment by SteveMacenski** [comment]
As a top level summary, my concerns here are:
- Algorithms around update with IDLE/running
- Default rate
- Not any tests for the speed controller node 
- No PR to add to BT nodes list and migration guides in the navigation.ros.org repo.

---

> On the first update, the history should be [1] and cumulated is also 1, so the returned value will be 1. On the second update (assuming the stamps are within the duration limit), the history will be [1, 2], cumulated will be 1 + 2 = 3 and returned value will be 3/2 = 1.5. On the third update, history is [1, 2, 3], cumulated is 3 + 3 = 6, and returned value is 6/3 = 2.

I looked over your code again and I think I agree, I thought `odom_cumulate_` was the rolling current odometry state. I think I realize now that its just the sum of the quantities and you don't actually divide them until when setting `vel_smooth_` so `odom_cumulate` doesn't have a division in it. 

> Let me explain the behavior of this node more clearly (I'll update the description as well). ... If the planning subtree was executed, SpeedController returns its status based on the status of the planning subtree

I think you just explained your own issue. After the replanning occurs, it returns the success / failure state and then its goes back to the `IDLE` state. Then your logic will trigger to force it to be the average update rate since `status() == BT::NodeStatus::IDLE`. Therefore, this speed controller will not actually update its rate between replans to the speed. The only time the rate should be reset to the average is on the very first tick when the BT starts up and the very first tick after a cancel or new navigation request (but should **not** when preempted - which this should do that correctly, right?). 

On your graph, I think that's totally fine. Its OK if it replans super quickly then pops back down. If you do average rate rather than max rate, you'll still see the exact same spike, its just only half-way up the chart. I think its a better standing assumption on startup take the most nuclear condition possible to avoid edge cases at low speeds on new requests.

**Comment by naiveHobo** [comment]
> After the replanning occurs, it returns the success / failure state and then its goes back to the IDLE state. Then your logic will trigger to force it to be the average update rate since status() == BT::NodeStatus::IDLE. Therefore, this speed controller will not actually update its rate between replans to the speed.

The node does not go back to IDLE after success / failure. It doesn't even go back to IDLE on preemption actually. The only time the node is IDLE is on startup and after a goal is reached. With that, I believe the logic of the node should be fine.

The only thing left to discuss is the default rate on IDLE. I'm fine with setting it to the max replanning rate value on IDLE. I agree that with the average rate, we will still see the exact same spike but it gives more time for the bot to gain speed before triggering the next replan which would eliminate that sharp fall on the third arrow in the above graph. But again, these values are dependent on the user defined `max_rate` and `min_rate` parameters so we can't say it will perform the same way in all cases.

> I think its a better standing assumption on startup take the most nuclear condition possible to avoid edge cases at low speeds on new requests.

I agree, that makes for a simpler assumption. I'll change the default value on IDLE.

File: nav2_behavior_tree/test/plugins/decorator/test_speed_controller.cpp [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+using namespace std::chrono;  // NOLINT
+using namespace std::chrono_literals;  // NOLINT
+
+class SpeedControllerTestFixture : public ::testing::Test
+{
+public:
+  static void SetUpTestCase()
+  {
+    transform_handler_ = new nav2_behavior_tree::TransformHandler();
```
Comment (SteveMacenski) [inline nav2_behavior_tree/test/plugins/decorator/test_speed_controller.cpp:37]: While I totally support the use of raw pointers when it makes sense, in this case it might be best to use smart pointers for consistency.

File: nav2_behavior_tree/test/plugins/decorator/test_speed_controller.cpp [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+  static nav2_behavior_tree::SpeedController * node_;
+  static nav2_behavior_tree::DummyNode * dummy_node_;
+};
+
+nav2_behavior_tree::TransformHandler * SpeedControllerTestFixture::transform_handler_ = nullptr;
+BT::NodeConfiguration * SpeedControllerTestFixture::config_ = nullptr;
+nav2_behavior_tree::SpeedController * SpeedControllerTestFixture::node_ = nullptr;
+nav2_behavior_tree::DummyNode * SpeedControllerTestFixture::dummy_node_ = nullptr;
+
+TEST_F(SpeedControllerTestFixture, test_behavior)
```
Comment (SteveMacenski) [inline nav2_behavior_tree/test/plugins/decorator/test_speed_controller.cpp:94]: A comment here about what this is doing would be good. Because I was reviewing the PR I'm familiar with it enough to pick up on the fact that those sleep durations are specifically meaningful relative the to the odometry speed, but that might not be immediately clear.

**Comment by SteveMacenski** [comment]
> The node does not go back to IDLE after success / failure. It doesn't even go back to IDLE on preemption actually. The only time the node is IDLE is on startup and after a goal is reached. With that, I believe the logic of the node should be fine.

Doesnt [this](https://github.com/ros-planning/navigation2/pull/1744/files#diff-d1ed5d13b0f23c40717ce3d25d17232aR108) show that it goes back to idle after success?

If that's actually true, then that would fix things for me. But I don't think that's always true. If a `halt` is propagated in the subtree because it was preempted (or other BT logic) while running, then I think that would set all nodes as `IDLE` that were `RUNNING` in that subtree, including this one.

**Comment by naiveHobo** [comment]
> Doesnt this show that it goes back to idle after success?

That's the child node which gets set to IDLE based on [this](https://github.com/BehaviorTree/BehaviorTree.CPP/blob/561f4fa86b45c64c9bb2fd3273f81363e6fd222f/src/decorator_node.cpp#L73-L76)

The decorator itself doesn't get set to IDLE at any point other than when its parent sets it to IDLE.

**Comment by SteveMacenski** [comment]
Agreed, but when a control node above it [asks it to halt](https://github.com/BehaviorTree/BehaviorTree.CPP/blob/561f4fa86b45c64c9bb2fd3273f81363e6fd222f/src/control_node.cpp#L44), then it would be set to idle. That could happen, for instance, if another subtree to the left of this one has a IsNewGoal or some conditional triggering it to `halt` the others. So `IDLE` can be triggered by state changes in the BT, not just related to the first `tick` or on `cancel` or new navigation request.

**Comment by naiveHobo** [comment]
Ah, I understand what you're saying.

I believe there's no general way to differentiate between a halt initiated due to an overall success (for eg. when a goal is reached) or due to some other condition like the one you mentioned above.

One way to reset the period to maximum rate ONLY when a new goal is available is to follow the logic in [goal_updated_condition](https://github.com/ros-planning/navigation2/blob/master/nav2_behavior_tree/plugins/condition/goal_updated_condition.cpp) and reset the period to max value when a new goal is available. But this would also reset the period on preemption (which should not happen, right?).

Though, I don't see a problem in setting the period to the maximum replanning rate any time the node is IDLE. In general, I believe when the decorator node is IDLE, it will be due to a condition which would justify resetting the period (for eg the IsNewGoal conditional you mentioned in your example). I can't think of a scenario where a halt caused by some condition other than goal reached or goal canceled would want the period to use the last value instead of resetting.

I believe we'll have to venture into some custom base node territory to be able to differentiate between different halt calls which is something I would want to avoid.

**Comment by SteveMacenski** [comment]
> One way to reset the period to maximum rate ONLY when a new goal is available is to follow the logic in goal_updated_condition and reset the period to max value when a new goal is available. But this would also reset the period on preemption (which should not happen, right?).

Very good thought, yea we could do that. It would reset on preemption but that might* be OK. On preemption, we want to replan immediately, so the SpeedController should allow the planner to execute. 

> I believe when the decorator node is IDLE, it will be due to a condition which would justify resetting the period

I would in general agree with that statement. My fear for something like the behavior tree is to be as complete as possible because you never know how someone will choose to use this later. Maybe rather than throttling the planner by speed, they throttle the controller, the EKF, or some 3rd party non-navigation thing. Maybe that's not as warranted. 

Agreed, we're not making a new base node. I want to fix what we can to be general and whatever we can't be general about, we explicitly document.

**Comment by naiveHobo** [comment]
Just to recap, speed controller now sets the period to max only when there's a new goal. In all other cases, when the last period is exceeded, the new period is set based on the current smoothed speed.

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/speed_controller.hpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+public:
+  SpeedController(
+    const std::string & name,
+    const BT::NodeConfiguration & conf);
+
+  // Any BT node that accepts parameters must provide a requiredNodeParameters method
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<double>("min_rate", 0.1, "Minimum rate"),
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/speed_controller.hpp:42]: Because this has been sitting around for so long, we need to add this to our new docs. https://github.com/ros-planning/navigation.ros.org/pull/25 and https://github.com/ros-planning/navigation2/blob/master/doc/parameters/param_list.md

File: nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp [hunk 17]
```diff
@@ -57,34 +56,31 @@ class DistanceControllerTestFixture : public ::testing::Test
...
-    delete dummy_node_;
-    transform_handler_ = nullptr;
-    config_ = nullptr;
-    dummy_node_ = nullptr;
   }
 
   void SetUp()
   {
     node_ = new nav2_behavior_tree::DistanceController("distance_controller", *config_);
+    dummy_node_ = new nav2_behavior_tree::DummyNode();
```
Comment (SteveMacenski) [inline nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp:64]: This was never deleted

File: nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp [hunk 18]
```diff
@@ -57,34 +56,31 @@ class DistanceControllerTestFixture : public ::testing::Test
...
-    delete config_;
-    delete dummy_node_;
-    transform_handler_ = nullptr;
-    config_ = nullptr;
-    dummy_node_ = nullptr;
   }
 
   void SetUp()
   {
     node_ = new nav2_behavior_tree::DistanceController("distance_controller", *config_);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp:63]: Was this deleted?

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 19]
```diff
@@ -64,6 +65,7 @@ BtNavigator::BtNavigator()
   declare_parameter("transform_tolerance", rclcpp::ParameterValue(0.1));
   declare_parameter("global_frame", std::string("map"));
   declare_parameter("robot_base_frame", std::string("base_link"));
+  declare_parameter("odom_topic", std::string("odom"));
```
Comment (SteveMacenski) [inline nav2_bt_navigator/src/bt_navigator.cpp:68]: This param also needs to be added to the MD file in Nav2 and the website parameter file for the bt navigator.

File: nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp (hunk 18 above)
Comment (naiveHobo) [reply nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp:63]: There was a suggestion earlier about using smart pointers in these tests wherever possible. I added that to the speed controller test so I ended up changing the distance controller test as well.

Comment (SteveMacenski) [reply nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp:63]: This isn't a smart pointer though, this is a raw pointer if you use `new`. If you use `new` you must delete it or you're leaking memory. Making these all smart pointers I think is the best call, but honestly I don't really care because its a test, it just needs to be properly deleted so we don't leak at bare minimum.

Comment (naiveHobo) [reply nav2_behavior_tree/test/plugins/decorator/test_distance_controller.cpp:65]: Ah yes I misunderstood what you meant by deleted. Should be okay now.

**Comment by SteveMacenski** [comment]
Last thing is to add these doc changes against ros-planning/navigation.ros.org#25

**Comment by naiveHobo** [comment]
Should be good to go now @SteveMacenski

**Comment by SteveMacenski** [comment]
Awesome :100: super happy to have this in :-)
